File: apps/android/app/src/main/kotlin/dev/w3ctech/infralab/ui/qr/QrApproveViewModel.kt
package dev.w3ctech.infralab.ui.qr

import dev.w3ctech.infralab.data.AuthClient
import dev.w3ctech.infralab.di.ServiceLocator
import dev.w3ctech.infralab.ui.auth.AuthMessages
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/** Where the QR approval flow is in its lifecycle (drives the account-screen card). */
enum class QrApproveStatus { IDLE, APPROVING, SUCCESS, ERROR }

data class QrApproveUiState(
    val status: QrApproveStatus = QrApproveStatus.IDLE,
    /** Error copy when [status] is ERROR; success copy when SUCCESS; otherwise null. */
    val message: String? = null,
)

/**
 * Drives the native side of QR cross-device login: an already-authenticated user scans the QR the
 * web renders (its content is the public ticket id) and this approves it, binding the browser's
 * pending ticket to the current user. Mirrors iOS's `approveQrLogin` client call; scanning itself
 * lives in the UI (Google's code scanner), so this ViewModel stays framework-free and unit-testable.
 */
class QrApproveViewModel(private val auth: AuthClient) : ViewModel() {

    private val _state = MutableStateFlow(QrApproveUiState())
    val state: StateFlow<QrApproveUiState> = _state.asStateFlow()

    /** Approve the scanned [ticketId]. Blank input (a cancelled/empty scan) is ignored. */
    fun approve(ticketId: String) {
        val trimmed = ticketId.trim()
        if (trimmed.isEmpty()) return
        viewModelScope.launch {
            _state.update { it.copy(status = QrApproveStatus.APPROVING, message = null) }
            runCatching { auth.approveQrLogin(trimmed) }
                .onSuccess {
                    _state.update {
                        it.copy(status = QrApproveStatus.SUCCESS, message = "已确认,网页端即将登录。")
                    }
                }
                .onFailure { err ->
                    _state.update {
                        it.copy(status = QrApproveStatus.ERROR, message = AuthMessages.describe(err))
                    }
                }
        }
    }

    /** Return the card to its resting state (after the user dismisses a result). */
    fun reset() = _state.update { QrApproveUiState() }

    companion object {
        val Factory: ViewModelProvider.Factory = object : ViewModelProvider.Factory {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>): T =
                QrApproveViewModel(ServiceLocator.authClient()) as T
        }
    }
}

File: apps/macos/InfraLab/Qr/QrApproveViewModel.swift
import Foundation

/// Drives the native side of QR cross-device login: an already-authenticated user
/// scans the QR the web renders (its content is the public ticket id) and confirms
/// here, binding the browser's pending ticket to the current user. The Swift
/// counterpart of Android's `QrApproveViewModel`; scanning itself lives in the UI
/// (``QrScannerView``), so this stays framework-free and unit-testable.
@MainActor
final class QrApproveViewModel: ObservableObject {
    /// Where the approval flow is in its lifecycle (drives the account-sheet card).
    enum Phase: Equatable {
        case idle
        case approving
        case success(message: String)
        case failure(message: String)
    }

    @Published private(set) var phase: Phase = .idle
    /// A scanned ticket awaiting the user's explicit confirmation — non-nil drives
    /// the confirm alert. The approve request only fires from ``confirmPending()``.
    @Published private(set) var pendingTicketId: String?

    private let client: AuthClient

    init(client: AuthClient) {
        self.client = client
    }

    /// Stage the scanned payload for confirmation. Blank input (a cancelled or
    /// empty scan) is ignored, mirroring the Android client.
    func handleScan(_ payload: String) {
        let trimmed = payload.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        phase = .idle
        pendingTicketId = trimmed
    }

    /// The user confirmed on the alert: approve the ticket. Takes the id
    /// explicitly (the alert captures it at presentation) so approval can't race
    /// the alert-dismiss binding clearing ``pendingTicketId``.
    func approve(ticketId: String) async {
        pendingTicketId = nil
        phase = .approving
        do {
            try await client.approveQrLogin(ticketId: ticketId)
            phase = .success(message: "已确认，网页端即将登录。")
        } catch let error as AuthClientError {
            phase = .failure(message: error.displayMessage)
        } catch {
            phase = .failure(message: AuthCopy.Errors.network)
        }
    }

    /// The user dismissed the confirm alert without approving.
    func cancelPending() {
        pendingTicketId = nil
    }

    /// Return the card to its resting state (before starting another scan).
    func reset() {
        phase = .idle
        pendingTicketId = nil
    }
}
